internal/gitserver/client.go

```diff
@@ -614,6 +631,163 @@ var deadlineExceededCounter = promauto.NewCounter(prometheus.CounterOpts{
+		}
+
+		var response protocol.BatchLogResponse
+		if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
+			return err
+		}
+		logger.Log(log.Int("numResults", len(response.Results)))
+
+		for _, result := range response.Results {
+			var err error
+			if result.CommandError != "" {
+				err = errors.Newf(result.CommandError)
+			}
+
+			rawResult := RawBatchLogResult{
+				Stdout: result.CommandOutput,
+				Error:  err,
+			}
+			if err := callback(result.RepoCommit, rawResult); err != nil {
+				return errors.Wrap(err, "commitLogCallback")
+			}
+
+			numProcessed++
+		}
+
+		return nil
+	}
+
+	// Find the gitserver shard each of the repositories referenced in the input belong to.
+	// We'll use these maps in the following to construct batches of requests with the correct
+	// mapping of repositories to shards.
+
+	addrsByName := make(map[api.RepoName]string, len(opts.RepoCommits))
+	for _, repoCommit := range opts.RepoCommits {
+		if _, ok := addrsByName[repoCommit.Repo]; ok {
+			// Already set
+			continue
+		}
+
+		addr, err := c.AddrForRepo(ctx, repoCommit.Repo)
+		if err != nil {
+			return err
+		}
+
+		addrsByName[repoCommit.Repo] = addr
+	}
+
+	// Construct batches of requests keyed by the address of the server that will receive the batch.
+	// The results from gitserver will have to be re-interlaced before returning to the client, so we
+	// don't need to be particularly concerned about order here.
+
+	batches := make(map[string][]api.RepoCommit, len(addrsByName))
+	for _, repoCommit := range opts.RepoCommits {
+		addr := addrsByName[repoCommit.Repo]
+
+		batches[addr] = append(batches[addr], api.RepoCommit{
+			Repo:     repoCommit.Repo,
+			CommitID: repoCommit.CommitID,
+		})
+	}
+
+	// Perform each batch reqeuest concurrently, but only allow four reqeuests to be in-flight at a
+	// time. This operation returns partial results in the case of a malformed or missing repository
+	// or a bad commit reference, but does not attempt to return partial results when an entire shard
+	// is down. Any of these operations failing will cause an error to be returned from the entire
+	// BatchLog function.
+
+	sem := semaphore.NewWeighted(int64(4))
+	g, ctx := errgroup.WithContext(ctx)
+
+	for addr, repoCommits := range batches {
+		// avoid capturing loop variable below
+		addr, repoCommits := addr, repoCommits
+
+		if err := sem.Acquire(ctx, 1); err != nil {
+			return err
+		}
+
+		g.Go(func() (err error) {
+			defer sem.Release(1)
+
+			return performLogRequestToShard(addr, repoCommits)
+		})
+	}
+
+	return g.Wait()
+}
+
+func repoNamesFromRepoCommits(repoCommits []api.RepoCommit) []string {
+	repoNameMap := make(map[api.RepoName]struct{}, len(repoCommits))
+	for _, rc := range repoCommits {
+		repoNameMap[rc.Repo] = struct{}{}
+	}
+	repoNames := make([]string, 0, len(repoNameMap))
+	for repoName := range repoNameMap {
+		repoNames = append(repoNames, string(repoName))
+	}
+	sort.Strings(repoNames)
```

Comment:
```suggestion
var repoNames []string
for _, rc := range repoCommits {
	if _, ok := repoNameMap[rc.Repo]; !ok {
		repoNameMap[rc.Repo] = struct{}{}
		repoNames = append(repoNames, string(rc.Repo))  
    }
}
```
1. Determinism without sorting.
2. IIUC the common case ought to be that most of the commits are for the same repo, so using `var repoNames []string` instead of `make` (but you could change that if I'm wrong).
3. Fusing the loops avoids iterating twice.

Reply:
1/3. Can do that.
2. Common case might be all *different* repos, if we're talking about find ref operations.

Reply:
Updated, PTAL.